packages/lib/src/vis/scatter/ScatterPoints.tsx

```diff
@@ -71,28 +72,46 @@ function ScatterPoints(props: Props) {
     [onPointerOut]
   );
 
-  const [dataGeometry] = useState(() => new BufferGeometry());
+  const { length } = data;
+  const { abscissaScale, ordinateScale } = useVisCanvasContext();
   const invalidate = useThree((state) => state.invalidate);
 
-  const dataToColorScale = useDataToColorScale(
+  const indexToPosition = useIndexToPosition(
+    abscissas,
+    abscissaScale,
+    ordinates,
+    ordinateScale
+  );
+
+  const valueToColor = useValueToColor(
     scaleType,
     domain,
     colorMap,
     invertColorMap
   );
 
-  const { position, color } = useBufferAttributes(
-    abscissas,
-    ordinates,
-    data,
-    dataToColorScale
-  );
+  const dataGeometry = useMemo(() => {
+    const geometry = new BufferGeometry();
+    geometry.setAttribute('position', createBufferAttr(length, 3));
+    geometry.setAttribute('color', createBufferAttr(length, 3));
```

Comment:
We now initialise the buffer attributes and their internal typed arrays only once, and re-initialise them only when `data.length` changes. As long as the number of values doesn't change, we can reuse the same instances.

Note that we were already initialising the typed arrays to the exact lengths we need, so that hasn't changed. But that's not what we do in some other places, so `createBufferAttr` will prove useful I think.

Note also that I'm now using a `Float32BufferAttribute` even for `color` instead of a "normalized" `Uint8BufferAttribute`, for the reason explained in the PR description.